SI-6623 Avoid $iw wrappers in REPL

som-snytt opened the pull request:
Use a conventional import to bump context depth in REPL templates.

Code is still ordinarily wrapped in a `$read` object.

This is a step toward 6623-like transparency.

`retronym` takes the blame for this innovation.
`adriaanm` collaborated in its commission.
`somsnytt` batted clean-up.

Notice how the handles line up.

som-snytt committed:
SI-8935 intp.valueOfTerm works with any template

Now `valueOfTerm` drills down reflectively using
the `fullName` of the desired symbol.

It respects a prefix of static modules, but switches
to instance mirrors as required.

The target is an accessor on the last enclosing instance.

cherrypicked from https://github.com/scala/scala/pull/4311

Conflicts:

	src/repl/scala/tools/nsc/interpreter/IMain.scala

retronym committed:
SI-6623 Avoid $iw wrappers in REPL

Use a conventional import to bump context depth in REPL templates.

Code is still ordinarily wrapped in a `$read` object.

This is a step toward 6623-like transparency.

`retronym` takes the blame for this innovation.
`adriaanm` collaborated in its commission.
`somsnytt` batted clean-up.

retronym commented on src/repl/scala/tools/nsc/interpreter/$iw.scala:
```diff
@@ -0,0 +1,9 @@
+//
+// `foo` will bind to `baz.foo` in this example:
+//
+// { import bar.foo; import interpreseter.$iw; import baz.foo; foo }
```
my bad: s/interpreseter/interpreter/

som-snytt replied:
Crosstalk from presentation compiler completer.

retronym commented on src/repl/scala/tools/nsc/interpreter/PresentationCompilerCompleter.scala:
```diff
@@ -40,12 +40,22 @@ class PresentationCompilerCompleter(intp: IMain) extends Completion {
-      Candidates(cursor, "" :: printed :: Nil)
+      val printed = s"${showCode(tree)} // : ${ if (tpe != null) tpe.safeToString else "???" }"
+      //Candidates(cursor, "" :: printed :: Nil)
+      Candidates(cursor, "" :: printed.lines.toList)   // avoid ctl-J in JLine?
```
Re `^J`: https://github.com/scala/scala-dev/issues/145

But in this case, maybe we should just do a `println` here, rather than trying to get jline to render this.

som-snytt replied:
I did see the `^J` issue. You had already suggested not printing the slain message, but just start fresh. Here, too, it would be nice if I didn't have to backspace over the `//print`. Perhaps?

retronym commented on test/files/run/repl-classbased.check:
```diff
@@ -6,18 +6,26 @@ scala> class C { implicit val k: K = K("OK?"); override def toString = s"C($k)"
 
 scala> val k = 42
+<console>:13: warning: imported `k' is permanently hidden by definition of value k
+import c.k
```
Seems like we've removed the special case in the typechecker that used to suppress this warning.

som-snytt commented on test/scaladoc/scalacheck/IndexScriptTest.scala:
```diff
@@ -46,7 +46,8 @@ object Test extends Properties("IndexScript") {
           "scala.tools.nsc.doc.html",
-          "scala.tools.nsc.doc.html.page"
+          "scala.tools.nsc.doc.html.page",
+          "scala.tools.nsc.interpreter"     // incurred by Contexts.isRootImport
```
@retronym I haven't nailed this down, but cracking open the test was enough trauma for one evening.

som-snytt commented on test/files/run/repl-colon-type.check:
```diff
@@ -34,12 +34,12 @@ scala> :type lazy val f = 5
+ object $read in package $line13 where target is defined
   lazy val $result = f
-                                           ^
+                                   ^
```
is that caret for line 36?

retronym commented on src/repl/scala/tools/nsc/interpreter/$iw.scala:
```diff
@@ -0,0 +1,9 @@
+// `foo` will bind to `baz.foo` in this example:
+//
+// { import bar.foo; import interpreseter.$iw; import baz.foo; foo }
+class $iw
```
Would be nice if this has a self describing name, rather than the nod to nostalgia, but I couldn't think of a decent name when I hacked this together? 
- `importDepthControl`
- `increaseImportDepth`
- ``{``
- ...

dwijnand replied:
`increaseImportNesting`?

som-snytt replied:
I actually considered double-brace, for printing purposes.

som-snytt commented:
`valueOfTerm` broke again, in `run/imain.scala`. Also, this was not the expected message.
```
> partest --show-log test/files/run/imain.scala
[error] Expected non-whitespace character
[error] partest --show-log test/files/run/imain.scala
[error]                   ^
```

retronym commented:
@som-snytt I looked at the `valueOfTerm` problem a bit but didn't find enlightenment. Could be a bug in the way it uses reflection, or a bug in reflection itself. I would suggest deleting that method for now.

som-snytt commented:
@retronym I had an old fix for `valueOfTerm` at https://github.com/scala/scala/pull/4311/commits which seems to work. More LOC.

But it has broken at SI-4899 and SI-5854 and SI-8935, which should tell us something. 

paulp:

> Impressed at the amount of ticket traffic for an unadvertised internal
> method. All the more reason to work toward that support repl API.
> Don't worry, it'll come.

Which should also tell us something.

som-snytt commented:
So, another evil of XML syntax. ScriptedTest failed with an obscure message, but luckily `// show` worked in the test to elucidate.
```
object $read extends scala.AnyRef {
  def <init>() = {
    super.<init>;
    ()
  };
  val INSTANCE = new $read.<init>
}
[error] Test scala.tools.nsc.interpreter.ScriptedTest.SI$minus8422$u0020captured$u0020multi$u0020i$divo failed: javax.script.ScriptException: in XML literal: in XML content, please use '}}' to express '}'
```

som-snytt commented:
Jenkins, old boy, I must say, I appreciate your company on these dark late nights. Perhaps I don't say it enough. But you're quite the chap.

som-snytt commented:
A few failures still.  This is a problem:
```
scala> k // show
object $read extends scala.AnyRef {
  def <init>() = {
    super.<init>;
    ()
  };
  import _root_.scala.tools.nsc.interpreter.$u007B$u007B;
  import _root_.scala.tools.nsc.interpreter.$u007B$u007B;
  import $line5.$read.c;
  import _root_.scala.tools.nsc.interpreter.$u007B$u007B;
  import c.k;
  import _root_.scala.tools.nsc.interpreter.$u007B$u007B;
  import $line11.$read.k;
  val res2 = k
}
<console>:17: error: reference to k is ambiguous;
it is imported twice in the same scope by
import k
and import c.k
       k // show
       ^
```

som-snytt commented on test/files/run/t7747-repl.check:
```diff
@@ -280,7 +273,7 @@ object $read extends scala.AnyRef {
   val INSTANCE = new $read.<init>
 }
-res3: List[Product with Serializable] = List(BippyBups(), PuppyPups(), Bingo())
+res3: List[Serializable with Product] = List(BippyBups(), PuppyPups(), Bingo())
```
Spurious update.

som-snytt commented on test/files/run/t7747-repl.check:
```diff
@@ -289,7 +282,7 @@ scala> val a = Sum("A", "B")
 
 scala> def b(a: Sum): String = a match { case Sum(_, _) => "Found Sum" }
-b: (a: Sum)String
+b: (a: INSTANCE.Sum)String
```
INSTANCE in reported type is worked around, but a correct fix is needed.

som-snytt committed:
SI-2458 Make spec example live test

Synchronize the live test with the spec update,
which is trivial.

Also add a neg test showing that an imported name
remains ambiguous even if it resolves to the
definition in scope with which it is ambiguous.

som-snytt commented:
/rebuild 6d8c565

som-snytt commented:
@adriaanm I don't know if this is useful pre-Adriaanite.

som-snytt commented on src/partest-extras/scala/tools/partest/ReplTest.scala:
```diff
@@ -31,20 +31,17 @@ abstract class ReplTest extends DirectTest {
+          case s          => s
+        }
+      } else {
+        lines drop header.lines.size
```
oh hey, this is where you cleaned up that code a little bit.
